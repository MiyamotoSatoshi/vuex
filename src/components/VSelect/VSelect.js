require('../../stylus/components/_text-fields.styl')
require('../../stylus/components/_input-groups.styl')
require('../../stylus/components/_select.styl')

// Components
import VBtn from '../VBtn'
import VCard from '../VCard'
import VCheckbox from '../VCheckbox'
import VChip from '../VChip'
import {
  VList,
  VListTile,
  VListTileAction,
  VListTileContent,
  VListTileTitle
} from '../VList'
import VMenu from '../VMenu'

// Mixins
import Colorable from '../../mixins/colorable'
import Dependent from '../../mixins/dependent'
import Filterable from '../../mixins/filterable'
import Input from '../../mixins/input'
import Maskable from '../../mixins/maskable'

// Component level mixins
import Autocomplete from './mixins/select-autocomplete'
import Computed from './mixins/select-computed'
import Events from './mixins/select-events'
import Generators from './mixins/select-generators'
import Helpers from './mixins/select-helpers'
import Menu from './mixins/select-menu'
import Props from './mixins/select-props'
import Watchers from './mixins/select-watchers'

// Directives
import ClickOutside from '../../directives/click-outside'

export default {
  name: 'v-select',

  inheritAttrs: false,

  components: {
    VCard,
    VCheckbox,
    VChip,
    VList,
    VListTile,
    VListTileAction,
    VListTileContent,
    VListTileTitle,
    VMenu,
    VBtn
  },

  directives: {
    ClickOutside
  },

  mixins: [
    Autocomplete,
    Colorable,
    Dependent,
    Events,
    Filterable,
    Generators,
    Helpers,
    Input,
    Maskable,
    Menu,
    Props,
    Watchers,
    // Input and Computed both
    // contain isDirty props
    // last gets merged in
    Computed
  ],

  data () {
    return {
      cachedItems: this.cacheItems ? this.items : [],
      content: {},
      defaultColor: 'primary',
      inputValue: (this.multiple || this.tags) && !this.value ? [] : this.value,
      isBooted: false,
      lastItem: 20,
      lazySearch: null,
      isActive: false,
      menuIsActive: false,
      searchTimeout: null,
      selectedIndex: -1,
      selectedItems: [],
      shouldBreak: false
    }
  },

  mounted () {
    // If instance is being destroyed
    // do not run mounted functions
    if (this._isDestroyed) return

    // Evaluate the selected items immediately
    // to avoid a unnecessary label transition
    this.genSelectedItems()

    this.content = this.$refs.menu.$refs.content
  },

  beforeDestroy () {
    if (this.isBooted) {
      if (this.content) {
        this.content.removeEventListener('scroll', this.onScroll, false)
      }
    }
  },

  methods: {
    changeSelectedIndex (keyCode) {
      // backspace, left, right, delete
      if (![8, 37, 39, 46].includes(keyCode)) return

      const indexes = this.selectedItems.length - 1

      if (keyCode === 37) { // Left arrow
        this.selectedIndex = this.selectedIndex === -1
          ? indexes
          : this.selectedIndex - 1
      } else if (keyCode === 39) { // Right arrow
        this.selectedIndex = this.selectedIndex >= indexes
          ? -1
          : this.selectedIndex + 1
      } else if (this.selectedIndex === -1) {
        this.selectedIndex = indexes
        return
      }

      // backspace/delete
      if ([8, 46].includes(keyCode)) {
        const newIndex = this.selectedIndex === indexes
          ? this.selectedIndex - 1
          : this.selectedItems[this.selectedIndex + 1]
            ? this.selectedIndex
            : -1

        this.combobox
          ? this.inputValue = null
          : this.selectItem(this.selectedItems[this.selectedIndex])
        this.selectedIndex = newIndex
      }
    },
    filterDuplicates (arr) {
      const values = arr.map(this.getValue)
      return arr.filter((el, i) => i === values.indexOf(values[i]))
    },
    genDirectives () {
      return [{
        name: 'click-outside',
        value: e => {
          return (
            !!this.content &&
            !this.content.contains(e.target) &&
            !!this.$el &&
            !this.$el.contains(e.target)
          )
        }
      }]
    },
    genSelectedItems (val = this.inputValue) {
      // If we are using tags, don't filter results
      if (this.tags) return (this.selectedItems = val)

      // Combobox is the single version
      // of a taggable select element
      if (this.combobox) return (this.selectedItems = val ? [val] : [])

      let selectedItems = this.computedItems.filter(i => {
        if (!this.isMultiple) {
          return this.getValue(i) === this.getValue(val)
        } else {
          // Always return Boolean
          return this.findExistingItem(i) > -1
        }
      })

      if (!selectedItems.length &&
        val != null &&
        this.tags
      ) {
        selectedItems = Array.isArray(val) ? val : [val]
      }

      this.selectedItems = selectedItems
    },
    clearableCallback () {
      const inputValue = this.isMultiple ? [] : null

      this.inputValue = inputValue
      this.$emit('change', inputValue)
      this.genSelectedItems()

      // When input is cleared
      // reset search value and
      // re-focus the input
      setTimeout(() => {
        this.searchValue = null
        this.focusInput()
      }, 0)

      if (this.openOnClear) {
        setTimeout(this.showMenu, 50)
      }
    },
    onScroll () {
      if (!this.isActive) {
        requestAnimationFrame(() => (this.content.scrollTop = 0))
      } else {
        if (this.lastItem >= this.computedItems.length) return

        const showMoreItems = (
          this.content.scrollHeight -
          (this.content.scrollTop +
          this.content.clientHeight)
        ) < 200

        if (showMoreItems) {
          this.lastItem += 20
        }
      }
    },
    findExistingItem (item) {
      return this.inputValue.findIndex((i) => {
        const a = this.getValue(i)
        const b = this.getValue(item)

        if (a !== Object(a)) return a === b

        return this.compareObjects(a, b)
      })
    },
    selectItem (item) {
      if (!this.isMultiple) {
        this.inputValue = this.returnObject ? item : this.getValue(item)
        this.selectedItems = [item]
      } else {
        const selectedItems = []
        const inputValue = this.inputValue.slice()
        const i = this.findExistingItem(item)

        i !== -1 && inputValue.splice(i, 1) || inputValue.push(item)
        this.inputValue = inputValue.map((i) => {
          selectedItems.push(i)
          return this.returnObject ? i : this.getValue(i)
        })

        this.selectedItems = selectedItems
      }

      this.searchValue = !this.isMultiple &&
        !this.chips &&
        !this.$scopedSlots.selection
        ? this.getText(this.selectedItem)
        : null

      this.$emit('change', this.inputValue)

      // List tile will re-render, reset index to
      // maintain highlighting
      const savedIndex = this.getMenuIndex()
      this.resetMenuIndex()

      // After selecting an item
      // refocus the input and
      // reset the caret pos
      this.$nextTick(() => {
        this.focusInput()
        this.setCaretPosition(this.currentRange)

        requestAnimationFrame(() => {
          if (savedIndex > -1) {
            this.setMenuIndex(savedIndex)
          }
        })
      })
    }
  },

  render (h) {
    const data = {
      attrs: {
        tabindex: this.isAutocomplete || this.disabled ? -1 : this.tabindex,
        ...(this.isAutocomplete ? null : this.$attrs),
        role: this.isAutocomplete ? null : 'combobox'
      }
    }

    if (!this.isAutocomplete) {
      data.on = this.genListeners()
      data.directives = this.genDirectives()
    } else {
      data.on = {
        click: () => {
          if (this.disabled || this.readonly || this.isFocused) return

          // If the input is dirty,
          // the input is not targetable
          // so we must manually focus
          if (this.isDirty) {
            this.focus()
            this.$nextTick(this.focusInput)
          }
        }
      }
    }

    return this.genInputGroup([
      this.genSelectionsAndSearch(),
      this.genMenu()
    ], data, this.toggleMenu)
  }
}
