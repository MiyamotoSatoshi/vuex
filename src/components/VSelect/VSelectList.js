// Components
import VCard from '../VCard'
import VCheckbox from '../VCheckbox'
import VDivider from '../VDivider'
import VSubheader from '../VSubheader'
import {
  VList,
  VListTile,
  VListTileAction,
  VListTileContent,
  VListTileTitle
} from '../VList'

// Mixins
import Colorable from '../../mixins/colorable'
import Themeable from '../../mixins/themeable'

// Helpers
import {
  escapeHTML,
  getPropertyFromItem
} from '../../util/helpers'

export default {
  name: 'v-select-list',

  mixins: [
    Colorable,
    Themeable
  ],

  props: {
    action: Boolean,
    dense: Boolean,
    hideSelected: Boolean,
    items: {
      type: Array,
      default: () => []
    },
    itemAvatar: {
      type: [String, Array, Function],
      default: 'avatar'
    },
    itemDisabled: {
      type: [String, Array, Function],
      default: 'disabled'
    },
    itemText: {
      type: [String, Array, Function],
      default: 'text'
    },
    itemValue: {
      type: [String, Array, Function],
      default: 'value'
    },
    noDataText: String,
    noFilter: Boolean,
    searchInput: {
      default: null
    },
    selectedItems: {
      type: Array,
      default: () => []
    }
  },

  computed: {
    parsedItems () {
      return this.selectedItems.map(item => this.getValue(item))
    },
    tileActiveClass () {
      return Object.keys(this.addTextColorClassChecks()).join(' ')
    }
  },

  methods: {
    genAction (item, inputValue) {
      const data = {
        on: {
          click: e => {
            e.stopPropagation()
            this.$emit('select', item)
          }
        }
      }

      return this.$createElement(VListTileAction, data, [
        this.$createElement(VCheckbox, {
          props: {
            color: this.computedColor,
            inputValue
          }
        })
      ])
    },
    genDivider (props) {
      return this.$createElement(VDivider, { props })
    },
    genFilteredText (text) {
      text = (text || '').toString()

      if (!this.searchInput || this.noFilter) return escapeHTML(text)

      const { start, middle, end } = this.getMaskedCharacters(text)

      return `${escapeHTML(start)}${this.genHighlight(middle)}${escapeHTML(end)}`
    },
    genHeader (props) {
      return this.$createElement(VSubheader, { props }, props.header)
    },
    genHighlight (text) {
      return `<span class="v-list__tile__mask">${escapeHTML(text)}</span>`
    },
    getMaskedCharacters (text) {
      const searchInput = (this.searchInput || '').toString().toLowerCase()
      const index = text.toLowerCase().indexOf(searchInput)

      if (index < 0) return { start: '', middle: text, end: '' }

      const start = text.slice(0, index)
      const middle = text.slice(index, index + searchInput.length)
      const end = text.slice(index + searchInput.length)
      return { start, middle, end }
    },
    genTile (
      item,
      disabled = null,
      avatar = false,
      value = this.parsedItems.indexOf(
        this.getValue(item)
      ) !== -1
    ) {
      if (item === Object(item)) {
        avatar = this.getAvatar(item)
        disabled = disabled !== null
          ? disabled
          : this.getDisabled(item)
      }

      const tile = {
        on: {
          mousedown: e => {
            // Prevent onBlur from being called
            e.preventDefault()
          },
          click: () => disabled || this.$emit('select', item)
        },
        props: {
          activeClass: this.tileActiveClass,
          avatar,
          disabled,
          ripple: true,
          value
        }
      }

      if (!this.$scopedSlots.item) {
        return this.$createElement(VListTile, tile, [
          this.action && !this.hideSelected && this.items.length > 0
            ? this.genAction(item, value)
            : null,
          this.genTileContent(item)
        ])
      }

      const parent = this
      const scopedSlot = this.$scopedSlots.item({ parent, item, tile })

      return this.needsTile(scopedSlot)
        ? this.$createElement(VListTile, tile, [scopedSlot])
        : scopedSlot
    },
    genTileContent (item) {
      const innerHTML = this.genFilteredText(this.getText(item))

      return this.$createElement(VListTileContent,
        [this.$createElement(VListTileTitle, {
          domProps: { innerHTML }
        })]
      )
    },
    needsTile (tile) {
      return tile.componentOptions == null ||
        tile.componentOptions.Ctor.options.name !== 'v-list-tile'
    },
    getAvatar (item) {
      return Boolean(getPropertyFromItem(item, this.itemAvatar, false))
    },
    getDisabled (item) {
      return Boolean(getPropertyFromItem(item, this.itemDisabled, false))
    },
    getText (item) {
      return (getPropertyFromItem(item, this.itemText, item) || '').toString()
    },
    getValue (item) {
      return getPropertyFromItem(item, this.itemValue, this.getText(item))
    }
  },

  render () {
    const children = []
    for (const item of this.items) {
      if (this.hideSelected &&
        this.selectedItems.indexOf(item) > -1
      ) continue

      if (item.header) children.push(this.genHeader(item))
      else if (item.divider) children.push(this.genDivider(item))
      else children.push(this.genTile(item))
    }

    if (!children.length) {
      const noData = this.$slots['no-data']
      if (noData) {
        children.push(noData)
      } else {
        children.push(this.genTile(this.noDataText, true))
      }
    }

    return this.$createElement(VCard, {
      staticClass: 'v-select-list',
      props: {
        dark: this.dark,
        light: this.light
      }
    }, [
      this.$createElement(VList, {
        props: {
          dense: this.dense
        }
      }, children)
    ])
  }
}
