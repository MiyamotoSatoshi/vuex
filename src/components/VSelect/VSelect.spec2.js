import { test } from '~util/testing'
import VSelect from '~components/VSelect'

test('VSelect.vue', ({ mount }) => {
  const up = new Event('keydown')
  const down = new Event('keydown')
  const space = new Event('keydown')
  const enter = new Event('keydown')

  up.keyCode = 38
  down.keyCode = 40
  space.keyCode = 32
  enter.keyCode = 13

  // Inspired by https://github.com/vuetifyjs/vuetify/pull/1425 - Thanks @kevmo314
  it('should open the select when focused and enter, space, up or down are pressed', async () => {
    const wrapper = mount(VSelect, {
      attachToDocument: true,
      propsData: {
        items: ['foo', 'bar']
      }
    })

    wrapper.vm.focus()

    wrapper.vm.$el.dispatchEvent(up)
    expect(wrapper.data().isActive).toBe(true)
    wrapper.vm.isActive = false
    wrapper.vm.$el.dispatchEvent(down)
    expect(wrapper.data().isActive).toBe(true)
    wrapper.vm.isActive = false
    wrapper.vm.$el.dispatchEvent(space)
    await wrapper.vm.$nextTick()
    expect(wrapper.data().isActive).toBe(true)
    wrapper.vm.isActive = false
    await wrapper.vm.$nextTick()
    wrapper.vm.$el.dispatchEvent(enter)
    await wrapper.vm.$nextTick()
    expect(wrapper.data().isActive).toBe(true)
    expect('Application is missing <v-app> component.').toHaveBeenTipped()
  })
})
