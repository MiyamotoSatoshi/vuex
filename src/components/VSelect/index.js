import VSelect from './VSelect'

/* istanbul ignore next */
VSelect.install = function install (Vue) {
  Vue.component(VSelect.name, VSelect)
}

export default VSelect
