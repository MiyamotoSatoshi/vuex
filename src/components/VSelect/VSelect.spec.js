import Vue from 'vue'
import { test } from '@util/testing'
import VSelect from '@components/VSelect'
import {
  VListTile,
  VListTileTitle,
  VListTileContent
} from '@components/VList'

test('VSelect', ({ mount, compileToFunctions }) => {
  it('should return numeric 0', () => {
    const item = { value: 0, text: '0' }
    const wrapper = mount(VSelect, {
      propsData: {
        value: null,
        items: [item],
        multiple: true
      }
    })

    const change = jest.fn()
    wrapper.vm.$on('change', change)
    wrapper.vm.selectItem(item)

    expect(change).toBeCalledWith([0])
    expect('Unable to locate target [data-app]').toHaveBeenTipped()
  })

  it('should be in an error state', async () => {
    const wrapper = mount(VSelect, {
      attachToDocument: true,
      propsData: {
        value: null,
        items: [0, 1, 2],
        rules: [(v) => !!v || 'Required']
      }
    })

    wrapper.trigger('focus')
    await wrapper.vm.$nextTick()
    wrapper.vm.blur()
    await wrapper.vm.$nextTick()

    expect(wrapper.vm.hasError).toBe(true)
    expect('Unable to locate target [data-app]').toHaveBeenTipped()
  })

  it('should disable list items', () => {
    const wrapper = mount(VSelect, {
      attachToDocument: true,
      propsData: {
        items: [{
          text: 'item',
          disabled: true
        }]
      }
    })

    const item = wrapper.find('li')[0]

    expect(item.element.getAttribute('disabled')).toBe('disabled')
    expect('Unable to locate target [data-app]').toHaveBeenTipped()
  })

  it('should warn when using incorrect item together with segmented prop', async () => {
    const items = [
      { text: 'Hello', callback: () => {} },
      { text: 'Hello' }
    ]

    const wrapper = mount(VSelect, {
      propsData: {
        segmented: true,
        items
      }
    })

    wrapper.vm.inputValue = items[1]

    await wrapper.vm.$nextTick()

    expect('Unable to locate target [data-app]').toHaveBeenTipped()
    expect('items must contain both a text and callback property').toHaveBeenTipped()
  })

  it('should render buttons correctly when using items array with segmented prop', async () => {
    const items = [
      { text: 'Hello', callback: () => {} }
    ]

    const wrapper = mount(VSelect, {
      propsData: {
        segmented: true,
        items
      }
    })

    wrapper.vm.inputValue = items[0]

    await wrapper.vm.$nextTick()

    expect(wrapper.html()).toMatchSnapshot()
    expect('Unable to locate target [data-app]').toHaveBeenTipped()
  })

  it('should render buttons correctly when using slot with segmented prop', async () => {
    const items = [
      { text: 'Hello' }
    ]

    const vm = new Vue()
    const selection = props => vm.$createElement('div', [props.item])
    const component = Vue.component('test', {
      components: {
        VSelect
      },
      render (h) {
        return h('v-select', {
          props: {
            segmented: true,
            items
          },
          scopedSlots: {
            selection
          }
        })
      }
    })

    const wrapper = mount(component)

    wrapper.vm.$children[0].inputValue = items[0]

    await wrapper.vm.$nextTick()

    expect(wrapper.html()).toMatchSnapshot()
    expect('Unable to locate target [data-app]').toHaveBeenTipped()
  })

  it('should render v-select correctly when using v-list-tile in item scope slot', async () => {
    const items = Array.from({ length: 2 }, (x, i) => ({ value: i, text: `Text ${i}` }))

    const vm = new Vue({
      components: {
        VListTile
      }
    })
    const itemSlot = ({ item, tile }) => vm.$createElement('v-list-tile', {
      on: tile.on,
      props: tile.props,
      class: item.value % 2 === 0 ? '' : 'red lighten-1'
    }, [
      item.text
    ])
    const component = Vue.component('test', {
      components: {
        VSelect
      },
      render (h) {
        return h('v-select', {
          props: {
            items
          },
          scopedSlots: {
            item: itemSlot
          }
        })
      }
    })

    const wrapper = mount(component)

    wrapper.vm.$children[0].inputValue = items[0]

    await wrapper.vm.$nextTick()

    expect(wrapper.html()).toMatchSnapshot()
    expect('Unable to locate target [data-app]').toHaveBeenTipped()
  })

  it('should render v-select correctly when not using v-list-tile in item scope slot', async () => {
    const items = Array.from({ length: 2 }, (x, i) => ({ value: i, text: `Text ${i}` }))

    const vm = new Vue({
      components: {
        VListTileTitle,
        VListTileContent
      }
    })
    const itemSlot = ({ item }) => vm.$createElement('v-list-tile-content', {
      class: item.value % 2 === 0 ? '' : 'red lighten-1'
    }, [
      vm.$createElement('v-list-tile-title', item.value)
    ])
    const component = Vue.component('test', {
      components: {
        VSelect
      },
      render (h) {
        return h('v-select', {
          props: {
            items
          },
          scopedSlots: {
            item: itemSlot
          }
        })
      }
    })

    const wrapper = mount(component)

    wrapper.vm.$children[0].inputValue = items[0]

    await wrapper.vm.$nextTick()

    expect(wrapper.html()).toMatchSnapshot()
    expect('Unable to locate target [data-app]').toHaveBeenTipped()
  })

  it('should render v-select correctly when not using scope slot', async () => {
    const items = Array.from({ length: 2 }, (x, i) => ({ value: i, text: `Text ${i}` }))

    const component = Vue.component('test', {
      components: {
        VSelect
      },
      render (h) {
        return h('v-select', {
          props: {
            items
          }
        })
      }
    })

    const wrapper = mount(component)

    wrapper.vm.$children[0].inputValue = items[0]

    await wrapper.vm.$nextTick()

    expect(wrapper.html()).toMatchSnapshot()
    expect('Unable to locate target [data-app]').toHaveBeenTipped()
  })

  it('should not close menu when using multiple prop', async () => {
    const wrapper = mount(VSelect, {
      attachToDocument: true,
      propsData: {
        items: [1, 2, 3, 4],
        multiple: true
      }
    })

    const blur = jest.fn()
    wrapper.vm.$on('blur', blur)

    wrapper.trigger('click')
    wrapper.trigger('blur')

    await wrapper.vm.$nextTick()

    const item = wrapper.find('li')[0]
    item.trigger('click')

    await wrapper.vm.$nextTick()

    expect(blur).not.toBeCalled()
    expect(wrapper.vm.isActive).toBe(true)
    expect('Unable to locate target [data-app]').toHaveBeenTipped()
  })

  it('should render aria-hidden=true on arrow icon', async () => {
    const wrapper = mount(VSelect)

    const icon = wrapper.find('.input-group__append-icon')[0]
    expect(icon.hasAttribute('aria-hidden')).toBe(true)
    expect('Unable to locate target [data-app]').toHaveBeenTipped()
  })

  it('should display a default value', async () => {
    const wrapper = mount(VSelect, {
      propsData: {
        value: 'foo',
        items: ['foo']
      }
    })

    expect(wrapper.vm.selectedItems).toEqual(['foo'])
    expect('Unable to locate target [data-app]').toHaveBeenTipped()
  })

  it('should not display a default value that is not in items', async () => {
    const wrapper = mount(VSelect, {
      propsData: {
        value: 'foo',
        items: ['bar']
      }
    })

    expect(wrapper.vm.selectedItems).toHaveLength(0)
    expect('Unable to locate target [data-app]').toHaveBeenTipped()
  })

  it('should update the displayed value when items changes', async () => {
    const wrapper = mount(VSelect, {
      propsData: {
        value: 1,
        items: []
      }
    })

    wrapper.setProps({ items: [{ text: 'foo', value: 1 }] })
    expect(wrapper.vm.selectedItems).toContainEqual({ text: 'foo', value: 1 })

    expect('Unable to locate target [data-app]').toHaveBeenTipped()
  })

  it('should render select menu with content class', async () => {
    const items = ['abc']

    const wrapper = mount(VSelect, {
      propsData: {
        contentClass: 'menu-class',
        items
      }
    })

    const menu = wrapper.find('.menu__content')[0]
    expect(menu.element.classList).toContain('menu-class')
    expect('Unable to locate target [data-app]').toHaveBeenTipped()
  })

  it('should have deletable chips', async () => {
    const wrapper = mount(VSelect, {
      attachToDocument: true,
      propsData: {
        chips: true,
        deletableChips: true,
        tags: true,
        items: ['foo', 'bar'],
        value: ['foo']
      }
    })

    await wrapper.vm.$nextTick()
    const chip = wrapper.find('.chip')[0]

    expect(!!chip).toBe(true)

    expect('Unable to locate target [data-app]').toHaveBeenTipped()
  })

  it('should escape items in menu', async () => {
    const wrapper = mount(VSelect, {
      propsData: {
        autocomplete: true,
        items: ['<strong>foo</strong>']
      }
    })

    const tileTitle = wrapper.find('.list__tile__title')[0]
    expect(tileTitle.html()).toMatchSnapshot()

    wrapper.setProps({ searchInput: 'str' })
    await wrapper.vm.$nextTick()
    expect(tileTitle.html()).toMatchSnapshot()

    expect('Unable to locate target [data-app]').toHaveBeenTipped()
  })

  it('should have the proper nudge', async () => {
    const wrapper = mount(VSelect, {
      attachToDocument: true,
      propsData: {
        hideDetails: true,
        items: ['foo', 'bar']
      }
    })

    expect(wrapper.vm.nudgeTop).toBe(-18)

    wrapper.setProps({ autocomplete: true })

    expect(wrapper.vm.nudgeTop).toBe(0)

    wrapper.setProps({ autocomplete: false, overflow: true })

    expect(wrapper.vm.nudgeTop).toBe(2)

    wrapper.setProps({ auto: true, overflow: false })

    expect(wrapper.vm.nudgeTop).toBe(-18)

    wrapper.setProps({ auto: false, overflow: true, hideDetails: false })

    expect(wrapper.vm.nudgeTop).toBe(26)

    wrapper.setProps({ hideDetails: true })

    expect(wrapper.vm.nudgeTop).toBe(2)

    expect('Unable to locate target [data-app]').toHaveBeenTipped()
  })

  it('should use value comparator', async () => {
    const wrapper = mount(VSelect, {
      attachToDocument: true,
      propsData: {
        multiple: true,
        items: [
          {text: 'one', value: 1},
          {text: 'two', value: 2},
          {text: 'three', value: 3}
        ],
        itemText: 'text',
        itemValue: 'value',
        valueComparator: (a, b) => Math.round(a) === Math.round(b),
        value: [3.1]
      }
    })

    expect(wrapper.vm.selectedItems).toHaveLength(1)
    expect(wrapper.vm.selectedItems[0].value).toBe(3)
    expect('Unable to locate target [data-app]').toHaveBeenTipped()
  })

  it('should not open if readonly', async () => {
    const wrapper = mount(VSelect, {
      propsData: {
        readonly: true,
        items: ['foo', 'bar']
      }
    })

    wrapper.trigger('click')
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.menuIsActive).toBe(false)

    wrapper.find('.input-group__append-icon')[0].trigger('click')
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.menuIsActive).toBe(false)

    expect('Unable to locate target [data-app]').toHaveBeenTipped()
  })
})
