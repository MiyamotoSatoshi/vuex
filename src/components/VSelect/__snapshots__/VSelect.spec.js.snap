// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VSelect should escape items in menu 1`] = `

<div class="list__tile__title">
  &lt;strong&gt;foo&lt;/strong&gt;
</div>

`;

exports[`VSelect should escape items in menu 2`] = `

<div class="list__tile__title">
  &lt;
  <span class="list__tile__mask">
    str
  </span>
  ong&gt;foo&lt;/strong&gt;
</div>

`;

exports[`VSelect should render buttons correctly when using items array with segmented prop 1`] = `

<div tabindex="0"
     data-uid="34"
     role="combobox"
     class="input-group input-group--dirty input-group--append-icon input-group--text-field input-group--select input-group--segmented input-group--single-line primary--text"
>
  <div class="input-group__input">
    <div class="input-group__selections"
         style="overflow: hidden;"
    >
      <button type="button"
              class="btn btn--flat"
              data-ripple="true"
      >
        <div class="btn__content">
          Hello
        </div>
      </button>
      <input disabled="disabled"
             tabindex="-1"
             class="input-group--select__autocomplete"
             style="display: none;"
      >
    </div>
    <div class="menu"
         style="display: inline-block;"
    >
      <div class="menu__content menu__content--select  menu__content--dropdown"
           style="max-height: 300px; min-width: 0px; max-width: auto; top: 12px; left: 0px; z-index: 0; display: none;"
      >
        <div class="card"
             style="height: auto;"
             data-ripple="false"
        >
          <ul class="list">
            <li>
              <a class="list__tile list__tile--link list__tile--active primary--text"
                 data-ripple="true"
              >
                <div class="list__tile__content">
                  <div class="list__tile__title">
                    Hello
                  </div>
                </div>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <i aria-hidden="true"
       class="icon material-icons input-group__append-icon input-group__icon-cb"
    >
      arrow_drop_down
    </i>
  </div>
  <div class="input-group__details">
  </div>
</div>

`;

exports[`VSelect should render buttons correctly when using slot with segmented prop 1`] = `

<div tabindex="0"
     data-uid="44"
     role="combobox"
     class="input-group input-group--dirty input-group--append-icon input-group--text-field input-group--select input-group--segmented input-group--single-line primary--text"
>
  <div class="input-group__input">
    <div class="input-group__selections"
         style="overflow: hidden;"
    >
      <div>
        Hello
      </div>
      <input disabled="disabled"
             tabindex="-1"
             class="input-group--select__autocomplete"
             style="display: none;"
      >
    </div>
    <div class="menu"
         style="display: inline-block;"
    >
      <div class="menu__content menu__content--select  menu__content--dropdown"
           style="max-height: 300px; min-width: 0px; max-width: auto; top: 12px; left: 0px; z-index: 0; display: none;"
      >
        <div class="card"
             style="height: auto;"
             data-ripple="false"
        >
          <ul class="list">
            <li>
              <a class="list__tile list__tile--link list__tile--active primary--text"
                 data-ripple="true"
              >
                <div class="list__tile__content">
                  <div class="list__tile__title">
                    Hello
                  </div>
                </div>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <i aria-hidden="true"
       class="icon material-icons input-group__append-icon input-group__icon-cb"
    >
      arrow_drop_down
    </i>
  </div>
  <div class="input-group__details">
  </div>
</div>

`;

exports[`VSelect should render v-select correctly when not using scope slot 1`] = `

<div tabindex="0"
     data-uid="72"
     role="combobox"
     class="input-group input-group--dirty input-group--append-icon input-group--text-field input-group--select primary--text"
>
  <div class="input-group__input">
    <div class="input-group__selections"
         style="overflow: hidden;"
    >
      <div class="input-group__selections__comma">
        Text 0
      </div>
      <input disabled="disabled"
             tabindex="-1"
             class="input-group--select__autocomplete"
             style="display: none;"
      >
    </div>
    <div class="menu"
         style="display: inline-block;"
    >
      <div class="menu__content menu__content--select"
           style="max-height: 300px; min-width: 0px; max-width: auto; top: 18px; left: 0px; z-index: 0; display: none;"
      >
        <div class="card"
             style="height: auto;"
             data-ripple="false"
        >
          <ul class="list">
            <li>
              <a class="list__tile list__tile--link list__tile--active primary--text"
                 data-ripple="true"
              >
                <div class="list__tile__content">
                  <div class="list__tile__title">
                    Text 0
                  </div>
                </div>
              </a>
            </li>
            <li>
              <a class="list__tile list__tile--link"
                 data-ripple="true"
              >
                <div class="list__tile__content">
                  <div class="list__tile__title">
                    Text 1
                  </div>
                </div>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <i aria-hidden="true"
       class="icon material-icons input-group__append-icon input-group__icon-cb"
    >
      arrow_drop_down
    </i>
  </div>
  <div class="input-group__details">
  </div>
</div>

`;

exports[`VSelect should render v-select correctly when not using v-list-tile in item scope slot 1`] = `

<div tabindex="0"
     data-uid="63"
     role="combobox"
     class="input-group input-group--dirty input-group--append-icon input-group--text-field input-group--select primary--text"
>
  <div class="input-group__input">
    <div class="input-group__selections"
         style="overflow: hidden;"
    >
      <div class="input-group__selections__comma">
        Text 0
      </div>
      <input disabled="disabled"
             tabindex="-1"
             class="input-group--select__autocomplete"
             style="display: none;"
      >
    </div>
    <div class="menu"
         style="display: inline-block;"
    >
      <div class="menu__content menu__content--select"
           style="max-height: 300px; min-width: 0px; max-width: auto; top: 18px; left: 0px; z-index: 0; display: none;"
      >
        <div class="card"
             style="height: auto;"
             data-ripple="false"
        >
          <ul class="list">
            <li>
              <a class="list__tile list__tile--link list__tile--active primary--text"
                 data-ripple="true"
              >
                <div class="list__tile__content">
                  <div class="list__tile__title">
                    0
                  </div>
                </div>
              </a>
            </li>
            <li>
              <a class="list__tile list__tile--link"
                 data-ripple="true"
              >
                <div class="list__tile__content red lighten-1">
                  <div class="list__tile__title">
                    1
                  </div>
                </div>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <i aria-hidden="true"
       class="icon material-icons input-group__append-icon input-group__icon-cb"
    >
      arrow_drop_down
    </i>
  </div>
  <div class="input-group__details">
  </div>
</div>

`;

exports[`VSelect should render v-select correctly when using v-list-tile in item scope slot 1`] = `

<div tabindex="0"
     data-uid="53"
     role="combobox"
     class="input-group input-group--dirty input-group--append-icon input-group--text-field input-group--select primary--text"
>
  <div class="input-group__input">
    <div class="input-group__selections"
         style="overflow: hidden;"
    >
      <div class="input-group__selections__comma">
        Text 0
      </div>
      <input disabled="disabled"
             tabindex="-1"
             class="input-group--select__autocomplete"
             style="display: none;"
      >
    </div>
    <div class="menu"
         style="display: inline-block;"
    >
      <div class="menu__content menu__content--select"
           style="max-height: 300px; min-width: 0px; max-width: auto; top: 18px; left: 0px; z-index: 0; display: none;"
      >
        <div class="card"
             style="height: auto;"
             data-ripple="false"
        >
          <ul class="list">
            <li class>
              <a class="list__tile list__tile--link list__tile--active primary--text"
                 data-ripple="true"
              >
                Text 0
              </a>
            </li>
            <li class="red lighten-1">
              <a class="list__tile list__tile--link"
                 data-ripple="true"
              >
                Text 1
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <i aria-hidden="true"
       class="icon material-icons input-group__append-icon input-group__icon-cb"
    >
      arrow_drop_down
    </i>
  </div>
  <div class="input-group__details">
  </div>
</div>

`;
